src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -603,13 +594,30 @@ ModifyQuerySupported(Query *queryTree, Query *originalQuery, bool multiShardQuer
 		}
 	}
 
-	/* reject queries which include CommonTableExpr */
+	/* reject queries which include CommonTableExpr which aren't routable */
 	if (queryTree->cteList != NIL)
 	{
-		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-							 "common table expressions are not supported in distributed "
-							 "modifications",
-							 NULL, NULL);
+		ListCell *cteCell = NULL;
+
+		foreach(cteCell, queryTree->cteList)
+		{
+			CommonTableExpr *cte = (CommonTableExpr *) lfirst(cteCell);
+			Query *cteQuery = (Query *) cte->ctequery;
+			DeferredErrorMessage *cteError = NULL;
+
+			if (cteQuery->commandType != CMD_SELECT)
+			{
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									 "Router planner doesn't support non-select common table expressions.",
+									 NULL, NULL);
+			}
+
+			cteError = MultiRouterPlannableQuery(cteQuery);
+			if (cteError)
+			{
+				return cteError;
+			}
```

Comment:
[Added bail for SELECT FOR UPDATE.](https://github.com/citusdata/citus/pull/2778/commits/6619e3d0f772c902d5ff75ca0d24c1438a1a8882) Not sure what "specialised logic" you were referring to, as the isolation test I added passes even when SELECT FOR UPDATE isn't bailed on

Reply:
So woke up & think I can recognize where this would've been lacking: need to make sure that update tasks get put on the same connection as the SELECT FOR UPDATE task